File: src/beaker_notebook/subkernels/julia/procedures/fetch_state.jl
{# Renders Julia code that produces a canonical KernelStatePayload. -#}
{# Composed from registered reflectors via a series of isa-checks.  -#}
using JSON3
using DisplayAs

_excluded_local_names = Set([{{ excluded_local_names | map('tojson') | join(', ') }}])

{% for reflector in reflectors -%}
# reflector: {{ reflector.name }} (target_types={{ reflector.target_types | join(', ') }})
{% include "reflectors/" ~ reflector.name ~ ".jl" %}

{% endfor %}

function _beaker_dispatch_reflector(_name::String, _value)
{% for reflector in reflectors %}
{% for target in reflector.target_types %}
{% if target == "__module__" %}
    if isa(_value, Module)
        return {{ reflector.function_name }}(_name, _value)
    end
{% elif target == "__function__" %}
    if isa(_value, Function)
        return {{ reflector.function_name }}(_name, _value)
    end
{% elif target == "__type__" %}
    if isa(_value, Type)
        return {{ reflector.function_name }}(_name, _value)
    end
{% elif target == "__default__" %}
    # default handled below
{% else %}
    try
        if isa(_value, {{ target }})
            return {{ reflector.function_name }}(_name, _value)
        end
    catch
    end
{% endif %}
{% endfor %}
{% endfor %}
{% for reflector in reflectors %}{% if "__default__" in reflector.target_types %}    return {{ reflector.function_name }}(_name, _value)
{% endif %}{% endfor %}
end

_module_syms = filter(k -> !startswith(string(k), "_") && !(string(k) in _excluded_local_names), names(@__MODULE__; imported=true))

_local_names = Dict{String, String}()
_variables = Dict{String, Any}()

for _sym in _module_syms
    _name = string(_sym)
    _value = try
        getfield(@__MODULE__, _sym)
    catch
        continue
    end
    _contribution = try
        _beaker_dispatch_reflector(_name, _value)
    catch err
        Dict("tag" => "<reflector error: $(typeof(err))>")
    end
    _local_names[_name] = get(_contribution, "tag", "<unknown>")
    if haskey(_contribution, "description")
        _variables[_name] = _contribution["description"]
    end
end

JSON3.write(Dict(
    "local_names" => _local_names,
    "variables" => _variables,
)) |> DisplayAs.unlimited

File: src/beaker_notebook/subkernels/python/procedures/fetch_state.py
{# Renders the Python fetch_state code that produces a canonical            -#}
{# KernelStatePayload (local_names + variables). Composed from the           -#}
{# reflectors registered on the subkernel.                                   -#}
import inspect as _inspect
import json as _json

class _BeakerStateEncoder(_json.JSONEncoder):
    def default(self, o):
        try:
            return super().default(o)
        except Exception:
            return str(o)

_EXCLUDED_LOCAL_NAMES = {{ excluded_local_names | tojson }}

{% for reflector in reflectors -%}
# reflector: {{ reflector.name }} (target_types={{ reflector.target_types | join(', ') }})
{% include "reflectors/" ~ reflector.name ~ ".py" %}

{% endfor %}

_REFLECTORS = {
{%- for reflector in reflectors %}
{%- for target in reflector.target_types %}
    {{ target | tojson }}: {{ reflector.function_name }},
{%- endfor %}
{%- endfor %}
}

def _beaker_dispatch_reflector(_name, _value):
    if _inspect.ismodule(_value):
        return _REFLECTORS["__module__"](_name, _value)
    if _inspect.isclass(_value):
        return _REFLECTORS["__class__"](_name, _value)
    if callable(_value):
        return _REFLECTORS["__callable__"](_name, _value)
    for _cls in type(_value).__mro__:
        _key = f"{_cls.__module__}.{_cls.__qualname__}"
        if _key in _REFLECTORS:
            return _REFLECTORS[_key](_name, _value)
    return _REFLECTORS["__default__"](_name, _value)

_result = {"local_names": {}, "variables": {}}
for _name, _value in dict(locals()).items():
    if _name.startswith("_"):
        continue
    if _name in _EXCLUDED_LOCAL_NAMES:
        continue
    try:
        _contribution = _beaker_dispatch_reflector(_name, _value)
    except Exception as _err:
        _contribution = {"tag": f"<reflector error: {type(_err).__name__}>"}
    _result["local_names"][_name] = _contribution.get("tag", "<unknown>")
    _description = _contribution.get("description")
    if _description is not None:
        _result["variables"][_name] = _description

_result = _json.loads(_json.dumps(_result, cls=_BeakerStateEncoder))
_result
